### DISCUSSION TITLE
Is it possible to set a minum diff for discharge?

### DISCUSSION THREAD

**Discussion by dlindh** [sonnen-tools/node-red#7] [Ideas]
Hi! 

Love this project :)

Is it already possible to set a diff value for discharge? Say that I charged the battery for 0,85 SEK but I only would like it to discharge if the diff is atleast 0,30 SEK? IE the price is atleast 1,15 SEK. Right now my battery starts discharge during the not so expensive hours only to be empty during the peak hours :)

Thanks :)

**Comment by markusjnagel** [comment]
Hi Daniel,

glad you like it :-)

Currently (actually, I'm not sure _when exactly_ I implemented it) it works as follows (sorry if you already know that):

It calculates an average (either across the whole time period or just the most expensive 5 hours)
(shown in the screenshot at the end as "average all" and "average top 5")
Depending on your settings, it will take one of these and calculate a Max Price (that is, the average you chose minus the conversion loss).
Then you can give it a max that you're willing to pay (User MaxPrice).
It will take the lower of these two.

Based on these values (the _effective_ Max price for charging) it will look for cheap hours. If found, it will try to charge the battery up to a target SoC during these hours.

So far, so good.

When it comes to discharging, first you need to enable the discharge control in the "Sonnenbatterie Config" node:
![image](https://github.com/sonnen-tools/node-red/assets/33287465/2d761518-63df-4a58-98d2-28d6697e5495)

**_If enabled, it will only allow discharging of the battery if:_**

The average price of what has been charged in the last 24hours multiplied with the conversion loss is cheaper than what the current Tibber price is (you can see the calculated Battery price as "Battery Price (incl. loss)" - if that is cheaper than Tibber, it will allow the battery to discharge.
If Tibber is cheaper, it will block the battery from discharging (as we don't want to lose money).

That way, it "shaves off" the most expensive hours. An example is [here](https://github.com/sonnen-tools/node-red/blob/main/docs/calculation.md#average-five-five-most-expensive-hours )

Up until v0.0.3 (haven't released a "cleaned up" version yet, only v0.0.3-WIP is under [Releases](https://github.com/sonnen-tools/node-red/releases)) that also meant that the battery can't charge from your solar system, if there is any surplus - it is "blocked", won't discharge but also won't charge. The surplus will then be fed into the grid. Not optimal when the sun comes out.

With v0.0.3, I introduced a threshold setting in the "Cheapest Hours Config" node called "pvSurplusThreshold".
![image](https://github.com/sonnen-tools/node-red/assets/33287465/eee5b28e-c5ba-4693-be4f-6f5fdd9908c5)

If exceeded, it will "release" the battery from its chains and sets it to "self-consumption". That means, it can pick up any surplus you might have. (the threshold is to avoid constant switching between modes when solar production and consumption fluctuates, default is 50W).

Please also check the [release notes](https://github.com/sonnen-tools/node-red/blob/main/releasenotes.md) for updated functionality that you might be interested in.

![image](https://github.com/sonnen-tools/node-red/assets/33287465/ae740749-5edd-4299-8dbd-cef59ccd7fe9)

**Comment by markusjnagel** [comment]
@dlindh - with that long-winded answer above: Would that meet your needs (battery only discharging if calculated battery price is lower than Tibber's, effectively covering the most expensive hours)? 

Or would you prefer an alternative "offset" way (calculated battery price including conversion loss needs to be lower by *[fixed number]* )?

**Reply by dlindh** [reply]
My first night with it working so I might have misunderstood :) 

 I have a 20kWh battery and live in northen Sweden in 250 square meter house and and temp between -30C and  -15C for the last week and a heat pump that takes energy from the ground to warm the house. I consume those 20kWh in an average of around 4-6 hours. (Right now my daily consumption is between 100-200 kWh ) (and yes the current 20kWh is just the start)
What I see now is that the battery charged last night until 06.00 when it started supplying the house because of the increased price. 

If we look at the price in Sweden for the two days here. The battery starts discharge at 06.00 (due to more expensive then what was charged I guess :) ) But at 06.00 the price is 1,37 SEK and at 08.00 its increased to 1,93 SEK. Then it goes down a litte and a new peak at 2.00 SEK at 17.00. So I wonder if there is a setting that will force the battery to not  discharge until the price is atleast the cost of charging multplied by X. Does that makes sense? I would like to battery to charge at the cheapest (Which is does) But also discharge when its most expensive :)
Image from tibber.
![image](https://github.com/sonnen-tools/node-red/assets/2614783/49706f5e-54ab-4012-9707-314ed594a079)

**Comment by markusjnagel** [comment]
I see your problem. Currently, we don't have these huge differences over the day in Germany (and I don't have a heatpump (yet))....

So, here's an idea:

1) I would add a setting for the size of the SB (20kWh in your case)
2) I would add a setting for the maximum discharge power (I think that is also 4.5kWh for the SB20, but I know it's less for the older SB8, I think).

So, assuming maximum discharge, that would mean in your case 20kWh/4.5kWh = (roughly) 4.5 hours. Based on that (rounded to the next full integer) I have the number of hours it takes for the SB to discharge under full power. Roughly 5hours in your case.

3) ok, we have 5 hours it takes to discharge.

4) I would then pick today's most expensive 5 hours and allow discharging only during those 5 hours - if the Tibber price is higher than what the calculated battery price is (price at the time of charging * loss_factor).

What do you think?

**Reply by markusjnagel** [reply]
if you wanted to "stretch" that time window, you would set the maximum discharge power to a lower value (I actually don't _*set*_ a discharge power, the battery does that by itself if it is in free "self-consumption" mode.. This would only be used for calculation of the size of the time window).

**Reply by dlindh** [reply]
Sounds like a great change but don't break your back for my corner case :) It would be greatly usable though. 
The battery has a 10kW inverter  or 9,9 in reality so maximum discharge could be a good thing :)

Thanks!

**Reply by markusjnagel** [reply]
Well, thinking about it - it is not really a corner case. It is a valid case for anyone whose typical consumption patterns (in your case the heatpump in cold Viking weather ;-)) will drain the battery in a short time.
Depends on the battery capacity and use case, but in general: If the consumption pattern outweighs the battery capacity, you would want to focus on the really expensive hours instead of just saving a bit and then being out of juice when it really becomes expensive.

Sounds valid to me. It won't bother you once you have extended your SB deployment to 200kWh ;-)))

**Reply by markusjnagel** [reply]
oh, and: 

> The battery has a 10kW inverter or 9,9 in reality so maximum discharge could be a good thing :)

Oh, cool, I didn't know that the large one also had a larger inverter. Great. However, I don't control the discharge power. Actually, I don't "force" a discharge. I just set the battery to its default "self-consumption" mode, where it will monitor your consumption and align its discharge power to your consumption (so that, ideally, you would have net zero grid consumption). With a little lag, of course, but... that's what it does.

But, if you have a bigger inverter (that allows a higher charging power) for your battery, you could set the charge power to a higher value in the "SonnenBatterie Config" node and see if it charges faster.
![image](https://github.com/sonnen-tools/node-red/assets/33287465/30f215d7-3bf2-43c6-b62d-ddb20e737548)

I wasn't aware that there are inverters with higher power ratings from Sonnen.

**Reply by doncuco** [reply]
No, must be a performance. my 10 eco with 22kwh is also only able do 4600W in/out

**Reply by dlindh** [reply]
Yes its a performance, and to be exact its discharge 9,9 kW and charge 8 kW :)

**Reply by markusjnagel** [reply]
oh man, I wish I had one... apparently, it uses denser battery modules. I wanted to extend mine (classic SB10) to 20 kWh, but it seems for a regular SB, you need another casing of the same size to do that - as opposed to the SB10 Performance, that can be extended to 20kWh in the same case..... 
As you can see, there isn't much space for a second wardrobe-sized battery in my cellar.... **_sigh_** :disappointed: 
![image](https://github.com/sonnen-tools/node-red/assets/33287465/49946bcf-544d-44a7-9888-a15510148535)

**Reply by markusjnagel** [reply]
@dlindh: 
> to be exact its discharge 9,9 kW and charge 8 kW :)

So, if you want faster charging, you could increase the `sonnen_charge_watts` value to 8000 - if you want, that is.

**Reply by dlindh** [reply]
Nice setup! I have reserved quite a lot of room so I have room for 10 or something cabinets if that would ever be needed or usable. 
![image](https://github.com/sonnen-tools/node-red/assets/2614783/05f24e9a-a850-4215-b5ca-22637c1a4e6a)

Also have 19440 kW of solar panels on the roof (solaredge) so I need more then the current 20kWh
![image](https://github.com/sonnen-tools/node-red/assets/2614783/5ae0068e-18e5-486d-b659-ecf18c999977)

**Reply by markusjnagel** [reply]
Wow :heart: love it - can I move in? :rofl:

**Reply by dlindh** [reply]
Plenty of space in rural Sweden to settle. And plenty of fiber availability too ;)

**Comment by markusjnagel** [comment]
by the way @dlindh - I like the idea and have created an issue from this discussion (that means, I will implement it and then close the issue, once done). https://github.com/sonnen-tools/node-red/issues/8

**Reply by dlindh** [reply]
Great! Thanks :) You should have a send a beer button or something for this project ;)

**Reply by markusjnagel** [reply]
Haha, yeah maybe. :beers: However, if you feel you want to give something, don't give it to me.

I bet there are local charities that would benefit from the Krona equivalent of a pint :beer: 
Or, give to a charity that supports Ukrainian refugees :ukraine: 
Or - I'm a strong believer in Open Source and sharing, transparency, openness. We grow when we share. 
I mean, I need this for myself anyway, so I could have just done this and kept it to myself - but why? If it helps as-is or as a learning opportunity or just some snippets of it help another project: Great! 
And together, we can make it better - I never would have thought of your use-case, for example. 
Everybody wins. :partying_face: 

With that said, here's another organisation that I fully endorse https://fsfe.org/index.en.html - if you like what they do, you could donate a pint to them. I'm especially a fan of the "Public Money? Public Code!" campaign: https://fsfe.org/activities/publiccode/publiccode.en.html

**Reply by dlindh** [reply]
I will send them a tip :) Ukraine is already on my monthly support list :)

**Reply by markusjnagel** [reply]
![image](https://github.com/sonnen-tools/node-red/assets/33287465/f3012911-644c-470d-8b20-e5930d7ba414)
